modules/llama_cpp_plugin/notebooks/qwen.ipynb

```diff
@@ -0,0 +1,281 @@
+   "id": "41f70a67-729c-4f3b-810c-d273d455e49e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "user_prompt = \"孙悟空是谁?\"\n",
+    "# user_prompt = \"太阳为什么是黄色的?\"\n",
+    "\n",
+    "formatted_input_prompt = convert_history([[user_prompt, \"\"]])\n",
+    "\n",
+    "from transformers import AutoTokenizer\n",
+    "tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)\n",
+    "\n",
+    "initial_prompt_tokens = tokenizer(formatted_input_prompt, return_tensors=\"np\", **tokenizer_kwargs).input_ids"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3c523b75-4acf-46cc-abb3-ff312321ff3d",
+   "metadata": {},
+   "source": [
+    "The initial prompt tokens are fed through the model to provide the context for the response generation.  Note that the regular OpenVINO Python API is used to supply inputs and receive outputs:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fa601c84-33d4-44a3-a27f-83f08eebe407",
+   "metadata": {},
+   "source": [
+    "The response tokens are generated one-by-one, conditioned on all previous tokens (within a certain window) by the KV-cache internally maintained by `llama.cpp`, until an end-of-sequence token is encountered as output or the maximum generated token count is reached."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cdbf4d60-519d-4c1d-862f-3796c362154c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "sequence_length = len(initial_prompt_tokens[0])\n",
+    "position_ids = np.arange(0, sequence_length).reshape(initial_prompt_tokens.shape)\n",
+    "\n",
+    "output = ov_model({\"input_ids\": initial_prompt_tokens, \"position_ids\": position_ids})\n",
+    "logits = output[\"logits\"]\n",
+    "curr_token_ids = np.argmax(logits[:, -1, :], axis=1).reshape([1, 1])\n",
+    "\n",
+    "MAX_TOKENS_GENERATED = 256\n",
+    "STOP_TOKENS = [tokenizer(st, return_tensors=\"np\").input_ids[0][0] for st in model_configuration[\"stop_tokens\"]]\n",
+    "\n",
+    "curr_tokens_generated = 0\n",
+    "last_token_id = curr_token_ids[0][0]\n",
+    "\n",
+    "response_tokens = []\n",
+    "next_position_id = sequence_length - 1\n",
+    "\n",
+    "while (last_token_id not in STOP_TOKENS) and (curr_tokens_generated < MAX_TOKENS_GENERATED):    \n",
+    "    print(tokenizer.decode(last_token_id), end='')\n",
+    "    curr_tokens_generated += 1\n",
+    "    curr_position_ids = np.ndarray([1, 1], dtype=np.int64)\n",
+    "    curr_position_ids[0][0] = next_position_id    \n",
+    "    next_position_id += 1\n",
+    "    curr_generated_output = ov_model({\"input_ids\": curr_token_ids, \"position_ids\": curr_position_ids})\n",
+    "    curr_logits = curr_generated_output[\"logits\"]\n",
+    "    curr_token_ids = np.argmax(curr_logits[:, -1, :], axis=1).reshape([1, 1])\n",
+    "    last_token_id = curr_token_ids[0][0]\n",
+    "\n",
+    "ov_model.create_infer_request().reset_state()"
```

Comment:
Does all `InferRequest` share state and one can reset it like this, or does this line create a new request and not affect the original?

Reply:
The state is shared across all infer requests to the same model object.

Reply:
Is it a LLAMA_CPP plugin feature?

Maybe the code sample should create an explicit `InferRequest` object anyway because it does not align with the rest of the OpenVINO (where each request has its own state) and might cause confusion in the future.

Reply:
The `InferRequest` object is explicitly created at the line you highlighted, during the evaluation of the `ov_model.create_infer_request()` subexpression. On the second look it might be possible to have separate KV caches associated with separate infer requests, with associated effects on memory consumption scaling, but then I don't see how the syntax in line 248 should work in a stateful fashion.

Reply:
This leaves us with a contradiction - the `CompiledModel._infer_request` is technically private so no user code can be written to reset the state of the initial infer request, and in case the states are to be handled as you are describing, then there is no way to reset the state of inference calls done via `CompiledModel.__call__`. 

The `LLAMA_CPP` plugin [implementations](https://github.com/openvinotoolkit/openvino_contrib/blob/8f238178616b47fb11fc3df0df99baf9e0df397d/modules/llama_cpp_plugin/src/infer_request.cpp#L124) of `InferRequest` resolve this situation by, in effect, sharing the state across all infer requests to a given model.

Reply:
agree that this behavior is weird.

In OpenVINO each infer request contains its own state to allow multiple independent requests to the same compiled model. Can we fix the behavior?

Reply:
I can fix that, but how then to reset the state of the internal `_infer_request` that is used by the Python API's `CompiledModel.__call__`?

Reply:
@jiwaszki how ca we reset state of internal request?

Reply:
> @jiwaszki how ca we reset state of internal request?

I think there is no way to get infer request from Compiled model. `__call__` method was added as a simple method.

> Alternatively, we can use explicit InferRequest API

So better to use explicit InferRequest

Reply:
@akuporos does this mean that the `__call__` API for OV inference is incompatible with stateful execution and all OV docs and notebooks should be adjusted to explicitly state this?

This could be easily fixed by providing a getter for the internal infer request in the `CompiledModel` object, I think.

Reply:
> This could be easily fixed by providing a getter for the internal infer request in the CompiledModel object, I think.

`CompiledModel.reset_state` might be more consistent with the existing API. Otherwise, there is not much sense in a hidden `InferRequest` in the first place.

Reply:
@vshampor @apaniukov Yes `CompiledModel.reset_state` can be more consistent and **you can create request for such feature**. The `CompiledModel.__call__` was added months ago as simplified method to work with OV, not as a functional replacement of `InferRequest` because of some voices that APIs "should align" -- that "call" was the most it could be done, let's keep that in mind.

> there is not much sense in a hidden InferRequest in the first place.

It is/was, because simplified API is what the name states.

> providing a getter for the internal infer request

If you need access right now, `CompiledModel._infer_request` will work as WA. If this is None (i.e. no inference was called yet) just skip the call.

So Python API would be happy to extend CompiledModel interface if there is direct need, but require story in JIRA to back it up.

@akuporos giving back the voice to you.

Reply:
@jiwaszki in the context of stateful execution this is not a feature request, but a bug report against the new API. I will submit the corresponding ticket. The user does not care about the reasons for introducing this exact version of inference, they care about fulfilling their use case. I think I have illustrated the direct need in this notebook - also 
the semi-casual user, familiar with the semantics of other inferencing frameworks, will try the `__call__` API first since it is, as you've said, "aligned" with what the other inferencing frameworks provide. Having `CompiledModel.reset_state` is fine by me.

> If you need access right now, CompiledModel._infer_request will work as WA.
That is understood, but surely you can't be recommending coding against internal APIs in a user-facing example.

Reply:
Meanwhile I rewrote the code to use `InferRequest`s explicitly.